lib/frontend/src/switch.rs

```diff
@@ -0,0 +1,193 @@
+            cases: HashMap::new(),
+        }
+    }
+
+    pub fn set_entry(&mut self, index: EntryIndex, ebb: Ebb) {
+        assert!(self.cases.insert(index, ebb).is_none(), "Tried to set the same entry {} twice", index);
+    }
+
+    fn build_cases_tree(self) -> Vec<(EntryIndex, Vec<Ebb>)> {
+        println!("before: {:#?}", self.cases);
+        let mut cases = self.cases.into_iter().collect::<Vec<(_, _)>>();
+        cases.sort_by_key(|&(index, _)| index);
+
+        let mut cases_tree: Vec<Vec<(EntryIndex, Ebb)>> = vec![];
+        let mut last_index = None;
+        for (index, ebb) in cases {
+            if Some(index - 1) > last_index {
+                cases_tree.push(Vec::new());
+            }
+            cases_tree.last_mut().unwrap().push((index, ebb));
+            last_index = Some(index);
+        }
+
+        println!("after: {:#?}", cases_tree);
+
+        cases_tree
+            .into_iter()
+            .map(|cases| {
+                (
+                    cases[0].0,
+                    cases.iter().map(|&(_, ebb)| ebb).collect(),
+                )
+            })
+            .collect()
+    }
+
+    pub fn emit(self, bx: &mut FunctionBuilder, val: Value, otherwise: Ebb) {
+        let cases_tree = self.build_cases_tree();
+
+        // FIXME icmp(_imm) doesn't have encodings for i8 and i16 on x86(_64) yet
+        let (val, val_type) = match bx.func.dfg.value_type(val) {
+            types::I8 | types::I16 => {
+                (
+                    bx.ins().uextend(types::I32, val),
+                    types::I32,
+                )
+            }
+            val_type => (val, val_type)
+        };
+
+        let cases_and_jt_ebbs: Vec<Option<(EntryIndex, Ebb, Vec<Ebb>)>> = cases_tree
+            .into_iter()
+            .rev()
+            .map(|(first_index, ebbs)| {
+                if ebbs.len() == 1 {
+                    let is_good_val = bx.ins().icmp_imm(IntCC::Equal, val, first_index);
+                    bx.ins().brnz(is_good_val, ebbs[0], &[]);
+                    None
+                } else {
+                    let jt_ebb = bx.create_ebb();
+                    let is_good_val = bx.ins().icmp_imm(IntCC::UnsignedGreaterThanOrEqual, val, first_index);
+                    bx.ins().brnz(is_good_val, jt_ebb, &[]);
+                    Some((first_index, jt_ebb, ebbs))
+                }
+            })
+            .collect();
+
+        bx.ins().jump(otherwise, &[]);
+
+        for elem in cases_and_jt_ebbs.into_iter().rev() {
+            if let Some((first_index, jt_ebb, ebbs)) = elem {
+                let mut jt_data = JumpTableData::new();
+                for ebb in ebbs {
+                    jt_data.push_entry(ebb);
+                }
+                let jump_table = bx.create_jump_table(jt_data);
+
+                bx.switch_to_block(jt_ebb);
+                let first_index = bx.ins().iconst(val_type, first_index);
+                let discr = bx.ins().isub(val, first_index);
```

Comment:
Can this be an `isub_imm` rather than needing a separate `iconst`?

Reply:
`isub_imm` doesnt exist.

Reply:
Oops, yes, the reason there's no isub_imm is because you can use iadd_imm with a negative immediate.